org-code-javabuilder/lib/src/main/java/org/code/javabuilder/ProjectFileLoader.java

```diff
@@ -0,0 +1,66 @@
+package org.code.javabuilder;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.concurrent.Callable;
+
+/**
+ * Fetches an individual file from a student's project. This should be used with the
+ * {dashboardHost}/v3/files/{channelId}/{fileName} API. It has no permissions to use AWS S3. This
+ * should be used asynchronously so multiple files can be retrieved at once.
+ */
+public class ProjectFileLoader implements Callable<Boolean> {
+  private final ProjectFile projectFile;
+  private final String fileUrl;
+
+  /**
+   * @param projectFile The in-memory representation of the given file
+   * @param fileUrl The URL to use to fetch the file
+   */
+  public ProjectFileLoader(ProjectFile projectFile, String fileUrl) {
+    this.projectFile = projectFile;
+    this.fileUrl = fileUrl;
+  }
+
+  /**
+   * @return true if the load succeeds
+   * @throws UserFacingException If any errors occur while fetching the file.
+   */
+  @Override
+  public Boolean call() throws UserFacingException {
```

Comment:
I'm not sure if this makes the code more or less readable, but it's possible to make this a function in UserProjectFileManager and then call it using an anonymous lambda.  (I suggest this because this class seems very tightly coupled with UserProjectFileManager and that would be a way of making that more explicit.)

Reply:
✔️